L1Trigger/L1THGCal/test/HGCalBackendStage1ParameterExtractor.cc

```diff
@@ -0,0 +1,293 @@
+    if (lpgbts.size() == 0)
+      continue;
+
+    // only retrieve mapping for the tested fpga
+    uint32_t fpgaId = triggerGeometry_->getStage1FpgaFromModule(moduleId);
+    HGCalTriggerBackendDetId tc_fpga(fpgaId);
+    if (!(tc_fpga.isStage1FPGA()) || (tc_fpga.sector() != 0) || (tc_fpga.zside() < 0))
+      continue;
+    int fpgaLabel = tc_fpga.label();
+    if (fpgaLabel != the_fpga_)
+      continue;
+
+    // retrieve information to be saved
+    int triggerCellSubdet = 0;
+    int triggerCellLayer = 0;
+    int triggerCellUEta = 0;
+    int triggerCellVPhi = 0;
+    if (id.det() == DetId::HGCalHSc) {
+      HGCScintillatorDetId id_sc(triggercell_cells.first);
+      triggerCellSubdet = id_sc.subdet();
+      triggerCellLayer = id_sc.layer();
+      triggerCellUEta = id_sc.ietaAbs();
+      triggerCellVPhi = id_sc.iphi();
+    } else if (HFNoseTriggerDetId(triggercell_cells.first).det() == DetId::HGCalTrigger &&
+               HFNoseTriggerDetId(triggercell_cells.first).subdet() == HGCalTriggerSubdetector::HFNoseTrigger) {
+      HFNoseTriggerDetId id_nose_trig(triggercell_cells.first);
+      triggerCellSubdet = id_nose_trig.subdet();
+      triggerCellLayer = id_nose_trig.layer();
+      triggerCellUEta = id_nose_trig.triggerCellU();
+      triggerCellVPhi = id_nose_trig.triggerCellV();
+    } else {
+      HGCalTriggerDetId id_si_trig(triggercell_cells.first);
+      triggerCellSubdet = id_si_trig.subdet();
+      triggerCellLayer = id_si_trig.layer();
+      triggerCellUEta = id_si_trig.triggerCellU();
+      triggerCellVPhi = id_si_trig.triggerCellV();
+    }
+    GlobalPoint position = triggerGeometry_->getTriggerCellPosition(triggercell_cells.first);
+    float triggerCellX = position.x();
+    float triggerCellY = position.y();
+    float triggerCellZ = position.z();
+    float triggerCellEta = position.eta();
+    float triggerCellPhi = position.phi();
+    float triggerCellRoverZ = sqrt(triggerCellX * triggerCellX + triggerCellY * triggerCellY) / triggerCellZ;
+
+    // transform HGCalHSc TC coordinates to define a TC address in [0,47]
+    uint32_t tce = triggerCellUEta;
+    uint32_t tcp = triggerCellVPhi;
+    if (triggerCellSubdet == 10) {  // HGCalHSc
+      tcp = (triggerCellVPhi - 1) % 4;
+      if (triggerCellUEta <= 3) {
+        tce = triggerCellUEta;
+      } else if (triggerCellUEta <= 9) {
+        tce = triggerCellUEta - 4;
+      } else if (triggerCellUEta <= 13) {
+        tce = triggerCellUEta - 10;
+      } else if (triggerCellUEta <= 17) {
+        tce = triggerCellUEta - 14;
+      } else {
+        tce = triggerCellUEta - 18;
+      }
+    }
+
+    // attribute ID to TC according to subdetector
+    uint32_t tcaddress = 99;
+    if (triggerCellSubdet == 10) {  //HGCalHSc(10)
+      tcaddress = (int(tce) << 2) + int(tcp);
+    } else {  //HGCalHSiTrigger(2) or HGCalEE(1)
+      tcaddress = tc_coord_uv_.find(std::make_pair(tce, tcp))->second;
+    }
```

Comment:
The sci address extraction would be better in separate dedicated function

Reply:
done, function returns a tuple with tcaddress and ueta/vphi